### ISSUE TITLE
[BUG] Exporting with Deepness > 1 crashed Admin with Memory Leak

### ISSUE THREAD

**Issue by julianwagle** [Baboo7/strapi-plugin-import-export-entries#42]
**Describe the bug**

If your schema is large enough and interconnected enough, running export with Deepness not set to one will crash the admin with a memory leak. 

**To Reproduce**

Reproduction will be hard if you are not using in a company with a large database and a large schema. If you can get access to this, all you would have to do to reproduce is try exporting with the default settings.

**Expected behavior**
Throwing an error in a try block if a certain number of connections exist or if manyToMany relationships exist which could cause an infinite loop. May also be smart to have an option in the config which allows the dev team to set a default and permanent deepness. If I could just set to one there would be no issues.

**Comment by Baboo7** [comment]
Interesting, I think your issue is partly linked to [this one](https://github.com/Baboo7/strapi-plugin-import-export-entries/issues/36)

How many records are you trying to export?

**Comment by julianwagle** [comment]
With deepness set to one - only up to a million. That works fine. The issues come up with linking. For example, if author in linked to book and book is linked to genre and genre is linked to authors and books the whole think collapses from a memory leak really fast, even if your book selection only holds a few thousand books

**Comment by julianwagle** [comment]
I solved by just brute force removing the option from the interface. In the event that someone else needs to do the same, here is the new ExportModal.js page:
 
`import { Button } from '@strapi/design-system/Button';
import { Checkbox } from '@strapi/design-system/Checkbox';
import { ModalLayout, ModalBody, ModalHeader, ModalFooter } from '@strapi/design-system/ModalLayout';
import { Flex } from '@strapi/design-system/Flex';
import { Grid, GridItem } from '@strapi/design-system/Grid';
import { Loader } from '@strapi/design-system/Loader';
import { Portal } from '@strapi/design-system/Portal';
import { Select, Option } from '@strapi/design-system/Select';
import { Typography } from '@strapi/design-system/Typography';
import pick from 'lodash/pick';
import React, { useState } from 'react';
import { useLocation } from 'react-router-dom';
import qs from 'qs';

import './style.css';
import ExportProxy from '../../api/exportProxy';
import { useDownloadFile } from '../../hooks/useDownloadFile';
import { useSlug } from '../../hooks/useSlug';
import { dataFormatConfigs, dataFormats } from '../../utils/dataFormats';
import { Editor } from '../Editor/Editor';
import { useAlerts } from '../../hooks/useAlerts';
import { handleRequestErr } from '../../utils/error';
import { useI18n } from '../../hooks/useI18n';

export const ExportModal = ({ onClose }) => {
  const { i18n } = useI18n();
  const { search } = useLocation();
  const { downloadFile, withTimestamp } = useDownloadFile();
  const { slug } = useSlug();
  const { notify } = useAlerts();

  const [optionExportFormat, setOptionExportFormat] = useState(dataFormats.CSV);
  const [optionApplyFilters, setOptionApplyFilters] = useState(false);
  const [optionRelationsAsId, setOptionRelationsAsId] = useState(false);
  const [data, setData] = useState(null);
  const [fetchingData, setFetchingData] = useState(false);

  const getData = async () => {
    setFetchingData(true);
    try {
      const res = await ExportProxy.getByContentType({
        slug,
        search: qs.stringify(pick(qs.parse(search), ['filters', 'sort'])),
        applySearch: optionApplyFilters,
        exportFormat: optionExportFormat,
        relationsAsId: optionRelationsAsId,
        deepness: 1,
      });
      setData(res.data);
    } catch (err) {
      handleRequestErr(err, {
        403: () => notify(i18n('plugin.message.export.error.forbidden.title'), i18n('plugin.message.export.error.forbidden.message'), 'danger'),
        default: () => notify(i18n('plugin.message.export.error.unexpected.title'), i18n('plugin.message.export.error.unexpected.message'), 'danger'),
      });
    } finally {
      setFetchingData(false);
    }
  };

  const writeDataToFile = async () => {
    const config = dataFormatConfigs[optionExportFormat];
    if (!config) {
      throw new Error(`File extension ${optionExportFormat} not supported to export data.`);
    }

    const { fileExt, fileContentType } = config;
    const fileName = `export_${slug}.${fileExt}`.replaceAll(':', '-').replaceAll('--', '-');
    downloadFile(data, withTimestamp(fileName), `${fileContentType};charset=utf-8;`);
  };

  const copyToClipboard = () => {
    navigator.clipboard.writeText(data);
    notify('Copied', '', 'success');
  };

  const clearData = () => {
    setData(null);
  };

  return (
    <Portal>
      <ModalLayout onClose={onClose} labelledBy="title">
        <ModalHeader>
          <Typography fontWeight="bold" textColor="neutral800" as="h2" id="title">
            Export
          </Typography>
        </ModalHeader>
        <ModalBody className="plugin-ie-export_modal_body">
          {fetchingData && (
            <>
              <Flex justifyContent="center">
                <Loader>Fetching data...</Loader>
              </Flex>
            </>
          )}
          {!data && !fetchingData && (
            <>
              <Grid gap={8}>
                <GridItem col={12}>
                  <Select id="export-format" label="Export Format" required placeholder="Export Format" value={optionExportFormat} onChange={setOptionExportFormat}>
                    <Option value={dataFormats.CSV}>{dataFormats.CSV.toUpperCase()}</Option>
                    <Option value={dataFormats.JSON}>{dataFormats.JSON.toUpperCase()}</Option>
                  </Select>
                </GridItem>
              </Grid>

              <Flex direction="column" alignItems="start" gap="16px">
                <Typography fontWeight="bold" textColor="neutral800" as="h2">
                  Options
                </Typography>
                <Checkbox value={optionRelationsAsId} onValueChange={setOptionRelationsAsId}>
                  Export relations as id.
                </Checkbox>
                <Checkbox value={optionApplyFilters} onValueChange={setOptionApplyFilters}>
                  Apply filters and sort to exported data.
                </Checkbox>
              </Flex>
            </>
          )}
          {data && !fetchingData && (
            <>
              <Editor content={data} language={optionExportFormat} />
            </>
          )}
        </ModalBody>
        <ModalFooter
          startActions={
            <>
              {!!data && (
                <Button variant="tertiary" onClick={clearData}>
                  {i18n('plugin.cta.back-to-options')}
                </Button>
              )}
            </>
          }
          endActions={
            <>
              {!data && <Button onClick={getData}>{i18n('plugin.cta.get-data')}</Button>}
              {!!data && (
                <>
                  <Button variant="secondary" onClick={copyToClipboard}>
                    {i18n('plugin.cta.copy-to-clipboard')}
                  </Button>
                  <Button onClick={writeDataToFile}>{i18n('plugin.cta.download-file')}</Button>
                </>
              )}
            </>
          }
        />
      </ModalLayout>
    </Portal>
  );
};
`

**Comment by Baboo7** [comment]
Hey @julianwagle , FYI I just published version `1.15.0` where v2 of export/import has been implemented. It's a beta still but it showed more stable results than the v1 on the test data I used, and a reduction of 80% in the size of the exported file. Export v2 is accessible in the export modal by selecting the export format `JSON (v2)` / import v2 is detected automatically from the structure of the provided data file.

I would love to have your feedback on this new version. Feel free to join the [discord server](https://discord.gg/cGRJh7M7) of the project I just opened 😌
